### ISSUE TITLE
Can't issue a read using metadata API

### ISSUE THREAD

**Issue by wadewegner** [jsforce/jsforce#621]
Having an issue reading Connected App metadata.

If I try ...
```
conn.metadata.read('ConnectedApp', 'PrettySweet', (readErr, metadataResult) => {
  console.log(metadataResult);
});
```
... the result is `{}`. But it's there.

Strangely, if I create, then read, like the following ...
```
conn.metadata.create('ConnectedApp', metadata, (createErr, results) => {
    conn.metadata.read('ConnectedApp', connectedAppName, (readErr, metadataResult) => {
      console.log(metadataResult);
    });
});
```
... it appears to work.

I'm also not sure how to just list out all Connected Apps. Is there a `.read` w/o the full name? Or a `.list`?

Thoughts on what I'm doing wrong?

Basically, I just want to list out all the details for my Connected Apps or a specific Connected App.

Thanks!

**Comment by paustint** [comment]
While I am not exactly sure what is wrong with the read call, aside from maybe the fullname being incorrect - I have also had some small issues with this call.

There is a listMetadata call that is available, but is not included in the main documentation: https://jsforce.github.io/jsforce/doc/Metadata.html#list, but I think that you need to start with describeMetadata https://jsforce.github.io/jsforce/doc/Metadata.html#describe before calling list, as list requires the specific metadata type name (unless you already know that).

One sidenote: the read call only supports 10 api names at a time, so you will have to batch up your requests if you need more.

- SFDC describeMetadata() https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_describe.htm
- SFDC listMetadata() https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_listmetadata.htm
- SFDC readMetadata() https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_readMetadata.htm

Here is my workflow (**note**: this is in typescript so I have specific type definitions for data, but you should get the idea pretty easily):
#### Describe all meatadata
**Note:** this will fail if version is not explicitly provided, from my experience.)
```javascript
export function describeMetadata(conn, version: string = '37.0') {
  return new Promise((resolve, reject) => {
    conn.metadata.describe(version, (err, metadataResults: DescribeMetadataResult) => {
      if (err) { return reject(err); }
      resolve(metadataResults);
    });
  });
}
```
#### List metadata for a specific type 
**Note:** I have noticed really slow performance when including multiple names. I found better results to query individually by metadata type and in parallel
```javascript
interface ListMetadataQuery {
  type: string,
  folder?: string;
}

export function listMetadata(conn, types: string | string[], version: string = '39.0') {
  return new Promise((resolve, reject) => {
    let typesArr: string[] = <string[]>(_.isArray(types) ? types : [types]);
    const queries: ListMetadataQuery[] = typesArr.map(type => {return {type: type}});
    conn.metadata.list(queries, version, (err, fileProperties: FileProperties) => {
      if (err) { return reject(err); }
      resolve(fileProperties);
    });
  });
}
```

#### Describe metadata return example:
**Note:** You will need to use the xmlName **AND** the childXmlNames (if present) to get all metadata types
```javascript
[
  {
    "type": "InstalledPackage",
    "describeMetadata": {
      "directoryName": "installedPackages",
      "inFolder": false,
      "metaFile": false,
      "suffix": "installedPackage",
      "xmlName": "InstalledPackage"
    }
  },
  {
    "type": "CustomLabels",
    "describeMetadata": {
      "childXmlNames": [
        "CustomLabel"
      ],
      "directoryName": "labels",
      "inFolder": false,
      "metaFile": false,
      "suffix": "labels",
      "xmlName": "CustomLabels"
    }
  }
...
]
```

#### List metadata return example:
**Note** One of the types in the describe call returned an error in the list call 100% of the time in every org, but I can't recall which one it is. Just beware if you are wanting to query everything and are using Promise.All() and trying to get everything.
```javascript
[
      {
        "createdById": "00541000001kiEEAAY",
        "createdByName": "Austin Turner",
        "createdDate": "2017-01-16T14:56:50.000Z",
        "fileName": "components/SBQQ__tabs.component",
        "fullName": "SBQQ__tabs",
        "id": "09941000000crHfAAI",
        "lastModifiedById": "00541000001kiEEAAY",
        "lastModifiedByName": "Austin Turner",
        "lastModifiedDate": "2017-01-16T14:56:50.000Z",
        "manageableState": "installed",
        "namespacePrefix": "SBQQ",
        "type": "ApexComponent"
      },
      {
        "createdById": "00541000001kiEEAAY",
        "createdByName": "Austin Turner",
        "createdDate": "2017-01-16T14:56:46.000Z",
        "fileName": "components/SBQQ__duelingListBox.component",
        "fullName": "SBQQ__duelingListBox",
        "id": "09941000000crGsAAI",
        "lastModifiedById": "00541000001kiEEAAY",
        "lastModifiedByName": "Austin Turner",
        "lastModifiedDate": "2017-01-16T14:56:46.000Z",
        "manageableState": "installed",
        "namespacePrefix": "SBQQ",
        "type": "ApexComponent"
      },
      {
        "createdById": "00541000001kiEEAAY",
        "createdByName": "Austin Turner",
        "createdDate": "2017-01-16T14:56:45.000Z",
        "fileName": "components/SBQQ__datePicker.component",
        "fullName": "SBQQ__datePicker",
        "id": "09941000000crGlAAI",
        "lastModifiedById": "00541000001kiEEAAY",
        "lastModifiedByName": "Austin Turner",
        "lastModifiedDate": "2017-01-16T14:56:45.000Z",
        "manageableState": "installed",
        "namespacePrefix": "SBQQ",
        "type": "ApexComponent"
      }
...
]
```

**Comment by vscuorzo** [comment]
I just ran into a similar issue. I first used the list method to get all of the connected apps and put their names into an array (with namespace prepended) which I then passed to the read method as the fullNames value. This failed and returned an "unexpected exception" error with no useful details explaining what went wrong. Trying to run similar code from the JSForce docs page also failed despite being connected to the same org and hardcoding the name of a connected app that I knew was present and was being returned by the list method run from the same page. 

What I discovered is that there are 2 issues here. In terms of the documentation page failure, it works if the fullNames array has more than one element in it, even if it's a duplicate value. This seems to be because the response will be an array only if multiple values are returned but the code to show the response always expects an array. For my code failure, this turned out to be due to a mysterious connected app called CPQIntegrationUserApp that is invisible in the UI but does get returned by the list method. As a result, this was being added to my fullNames array and apparently that app can't be retrieved using the read method, causing the underlying Metadata API call to fail. I updated my code to ignore that connected app and it works as expected.

Here's what ended up working:

`// list the connected apps
orgConnection.metadata.list([{type: 'ConnectedApp', folder: null}], '55.0', function(err, metadata) {
	if(err){ return console.error('err', err); }
	
	let appNames = [];
	metadata.forEach((connectedApp) => {
		let appName = (connectedApp.namespacePrefix ? connectedApp.namespacePrefix + '__' : '') + connectedApp.fullName;

		if(connectedApp.fullName != 'CPQIntegrationUserApp'){
			appNames.push(appName);
		}
		else{
			console.log('ignoring BS salesforce connected app: ' + appName);
		}
	});

	// get connected app info
	orgConnection.metadata.read('ConnectedApp', appNames, function(err, metadata) {
		if(err){ return console.error('err', err); }

		if(metadata instanceof Array){
			// process your results as an array here
		}
	});
});`
